io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs

```diff
@@ -202,140 +225,175 @@ impl<'n> Nexus<'n> {
+                        .map(
+                            |j| {
+                                if j.src_uri == src_uri {
+                                    Some(j)
+                                } else {
+                                    None
+                                }
+                            },
+                        )
+                        .flatten()
+                })
+                .collect()
+        }
     }
 
-    /// Return rebuild job associated with the dest child name.
-    /// Return error if no rebuild job associated with it.
-    fn get_rebuild_job<'a>(
+    /// Returns rebuild job associated with the destination child URI.
+    /// Returns error if no rebuild job associated with it.
+    pub fn rebuild_job(
         &self,
-        name: &'a str,
-    ) -> Result<&'a mut RebuildJob, Error> {
-        let job = RebuildJob::lookup(name).context(RebuildJobNotFound {
-            child: name.to_owned(),
-            name: self.name.clone(),
-        })?;
+        dst_child_uri: &str,
+    ) -> Result<&RebuildJob<'n>, Error> {
+        self.lookup_child(dst_child_uri)
+            .map(|c| c.rebuild_job())
+            .flatten()
+            .ok_or_else(|| Error::RebuildJobNotFound {
+                child: dst_child_uri.to_owned(),
+                name: self.name.clone(),
+            })
+    }
 
-        assert_eq!(job.nexus, self.name);
-        Ok(job)
+    /// Returns rebuild job associated with the destination child URI.
+    /// Returns error if no rebuild job associated with it.
+    pub fn rebuild_job_mut(
+        self: Pin<&mut Self>,
+        dst_child_uri: &str,
+    ) -> Result<&mut RebuildJob<'n>, Error> {
+        let name = self.name.clone();
+        self.lookup_child_mut(dst_child_uri)
+            .map(|c| c.rebuild_job_mut())
+            .flatten()
+            .ok_or_else(move || Error::RebuildJobNotFound {
+                child: dst_child_uri.to_owned(),
+                name,
+            })
+    }
+
+    /// Returns number of rebuild jobs on all children.
+    pub fn count_rebuild_jobs(&self) -> usize {
+        self.children_iter().fold(0, |acc, c| {
+            if c.rebuild_job().is_some() {
+                acc + 1
+            } else {
+                acc
+            }
+        })
     }
 
     /// On rebuild job completion it updates the child and the nexus
     /// based on the rebuild job's final state
-    async fn on_rebuild_complete_job(
+    async fn on_rebuild_update(
         mut self: Pin<&mut Self>,
-        job: &RebuildJob,
+        child_uri: &str,
     ) -> Result<(), Error> {
-        let recovering_child =
-            self.as_mut().get_child_by_name(&job.destination)?;
+        if !self.rebuild_job(child_uri)?.state().done() {
+            // Leave all states as they are
+            return Ok(());
+        }
+
+        let dst_child = self.as_mut().child_mut(child_uri)?;
+        let job = dst_child.remove_rebuild_job().unwrap();
```

Comment:
should this return an error? or maybe not since we're in the callback we know the job is there?

Reply:
The child is checked to have a job running in a line above.
But it won't hurt to add a check and a warning here instead of unwrap.